server/dns_server.mli

```diff
@@ -102,6 +102,9 @@ val handle_tsig : ?mac:Cstruct.t -> t -> Ptime.t -> Packet.t ->
 (** [handle_tsig ~mac t now packet buffer] verifies the tsig
     signature if present, returning the keyname, tsig, mac, and used key. *)
 
+type packet_callback = Packet.t -> Packet.t option
+(** [callback question] either returns a faux-handled DNS query [Some answer] or [None]. *)
```

Comment:
should its type be more constrained, i.e. `Packet.request -> Packet.reply option`? I'm struggling a bit what a good API would be -- and what to allow to clients of this API?

From the DNS server perspective, the invariant to satisfy is that each decoded request should be answered with a reply. This also means that stuff like ID in the header etc. should be good (so no client is confused) -- also fragmentation (including edns information), edns reply, ... would be nice to handle only at one point, wouldn't it?

What do you think? What is the information you really need in your use case? And would it simplify your code if you'd just reply with an answer (instead of a whole packet)?

Reply:
> should its type be more constrained, i.e. Packet.request -> Packet.reply option? I'm struggling a bit what a good API would be -- and what to allow to clients of this API?

That sounds like a good idea. What about Packet.Question.t -> Packet.reply? This would allow the client to see that name that's being queried.

Reply:
For me, `Question.t -> reply option` would be great. If that is all you need, perfect.

Reply:
Great! I'll change the commit to reflect this.